sql/core/src/test/scala/org/apache/spark/sql/FilterPushdownBenchmark.scala

```diff
@@ -0,0 +1,195 @@
+    val (keys, values) = pairs.unzip
+    val currentValues = keys.map(key => Try(spark.conf.get(key)).toOption)
+    (keys, values).zipped.foreach(spark.conf.set)
+    try f finally {
+      keys.zip(currentValues).foreach {
+        case (key, Some(value)) => spark.conf.set(key, value)
+        case (key, None) => spark.conf.unset(key)
+      }
+    }
+  }
+
+  private def prepareTable(dir: File, df: DataFrame): Unit = {
+    val dirORC = dir.getCanonicalPath + "/orc"
+    val dirParquet = dir.getCanonicalPath + "/parquet"
+
+    df.write.mode("overwrite").orc(dirORC)
+    df.write.mode("overwrite").parquet(dirParquet)
+
+    spark.read.orc(dirORC).createOrReplaceTempView("orcTable")
+    spark.read.parquet(dirParquet).createOrReplaceTempView("parquetTable")
+  }
+
+  def filterPushDownBenchmark(values: Int, width: Int, expr: String): Unit = {
+    val benchmark = new Benchmark(s"Filter Pushdown ($expr)", values)
+
+    withTempPath { dir =>
+      withTempTable("t1", "orcTable", "patquetTable") {
+        import spark.implicits._
+        val selectExpr = (1 to width).map(i => s"CAST(value AS STRING) c$i")
+        val df = spark.range(values).map(_ => Random.nextLong).selectExpr(selectExpr: _*)
+          .withColumn("id", monotonically_increasing_id())
+
+        df.createOrReplaceTempView("t1")
+        prepareTable(dir, spark.sql("SELECT * FROM t1"))
+
+        Seq(false, true).foreach { value =>
+          benchmark.addCase(s"Parquet Vectorized ${if (value) s"(Pushdown)" else ""}") { _ =>
+            withSQLConf(SQLConf.PARQUET_FILTER_PUSHDOWN_ENABLED.key -> s"$value") {
+              spark.sql(s"SELECT * FROM parquetTable WHERE $expr").collect()
+            }
+          }
+        }
+
+        Seq(false, true).foreach { value =>
+          benchmark.addCase(s"Native ORC Vectorized ${if (value) s"(Pushdown)" else ""}") { _ =>
+            withSQLConf(SQLConf.ORC_FILTER_PUSHDOWN_ENABLED.key -> s"$value") {
+              spark.sql(s"SELECT * FROM orcTable WHERE $expr").collect()
+            }
+          }
+        }
+
+        // Positive cases: Select one or no rows
+        /*
+        Java HotSpot(TM) 64-Bit Server VM 1.8.0_152-b16 on Mac OS X 10.13.2
+        Intel(R) Core(TM) i7-4770HQ CPU @ 2.20GHz
+
+        Filter Pushdown (id = 0):                Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
+        ------------------------------------------------------------------------------------------------
+        Parquet Vectorized                            2267 / 2287          0.5        2162.0       1.0X
+        Parquet Vectorized (Pushdown)                  735 /  803          1.4         701.1       3.1X
+        Native ORC Vectorized                         1708 / 1718          0.6        1629.1       1.3X
+        Native ORC Vectorized (Pushdown)                83 /   88         12.7          79.0      27.4X
+
+        Filter Pushdown (id == 0):               Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
```

Comment:
what's the difference between `id = 0` and `id == 0`? Do you want `id <=> 0`?

Reply:
Oops. Yes. `<=>`.